File: golang/internal/infrastructure/repository/portfolio_repository_impl.go
package repository

import (
	"sync"

	apprepository "folio/codinginterview/internal/application/repository"
	"folio/codinginterview/internal/domain"
)

type PortfolioRepositoryImpl struct {
	mu        sync.RWMutex
	portfolio domain.Portfolio
}

func NewPortfolioRepositoryImpl() apprepository.PortfolioRepository {
	return &PortfolioRepositoryImpl{
		portfolio: domain.MustInitialPortfolio(),
	}
}

func (r *PortfolioRepositoryImpl) Get() (domain.Portfolio, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.portfolio, nil
}

func (r *PortfolioRepositoryImpl) Update(portfolio domain.Portfolio) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.portfolio = portfolio
	return nil
}

File: typescript/src/infrastructure/repository/portfolioRepositoryImpl.ts
import { PortfolioRepository } from "../../application/repository/portfolioRepository";
import { AppConstants } from "../../domain/appConstants";
import { Portfolio } from "../../domain/stock";

export class PortfolioRepositoryImpl implements PortfolioRepository {
  private portfolio: Portfolio = AppConstants.initialPortfolio;

  async get(): Promise<Portfolio> {
    return this.portfolio;
  }

  async update(portfolio: Portfolio): Promise<void> {
    this.portfolio = portfolio;
  }
}
